Refactored parallel::CellWeights.

marcfehling opened the pull request:
We generalized the interface of the `parallel::CellWeights` class in such a way that we introduced static member functions. Those let the user manage connecting and disconnecting callback functions manually if desired without instantiating an object of this class.

The legacy interface has been deprecated. Corresponding tests have been updated, revealing a bug in `hp::DoFHandler`: All its internal data structures need to be cleared when calling `initialize()`.

Part of #7515.

@bangerth -- FYI

marcfehling commented:
When refactoring this class, we actually didn't use the feature introduced in #8672. It may be useful in the future, but is unused now.

Shall we leave it in or remove it?

bangerth commented on include/deal.II/distributed/cell_weights.h:
```diff
@@ -38,80 +38,178 @@ namespace parallel
-   * The returned value has to be an unsigned integer and is thus limited by
-   * some large number. It is interpreted as the additional computational load
-   * of each cell. See Triangulation::Signals::cell_weight for a discussion on
-   * this topic.
```
Would it be useful to preserve these four lines somehow, or at least link to a place where it is explained how the weighting works?

marcfehling replied:
Are you okay with moving the explanation to the `WeightingFunction` alias (as you suggested below) and leaving a `@note` referring to `Triangulation::Signals::cell_weight` in the general class documentation?

bangerth commented on include/deal.II/distributed/cell_weights.h:
```diff
@@ -38,80 +38,178 @@ namespace parallel
    * latter, an assertion will be triggered in the weight_callback() function.
-   * It is recommended to create a separate object in this case and to destroy
-   * the previous one.
+   * Use CellWeights::reinit() to update dependencies.
```
It's not clear to me what "dependencies" you refer to. Can you elaborate?

bangerth commented on include/deal.II/distributed/cell_weights.h:
```diff
@@ -38,80 +38,178 @@ namespace parallel
   public:
+    /**
+     * An alias that defines the characteristics of a function that can be used
+     * for weighting cells during load balancing.
```
```suggestion
     * for weighting cells during load balancing. Such weighting functions
     * take as arguments a finite element and an iterator to a cell
     * on which this element is to be used. They return an
     * unsigned integer (the weight).
```

bangerth commented on include/deal.II/distributed/cell_weights.h:
```diff
@@ -38,80 +38,178 @@ namespace parallel
+     * An alias that defines the characteristics of a function that can be used
+     * for weighting cells during load balancing.
+     */
+    using weighting_function = std::function<unsigned int(
```
Can we call it `WeightingFunction` to stick with typical `CameCase` for types?

bangerth commented on include/deal.II/distributed/cell_weights.h:
```diff
@@ -38,80 +38,178 @@ namespace parallel
+     */
+    using weighting_function = std::function<unsigned int(
+      const FiniteElement<dim, spacedim> &,
+      const typename hp::DoFHandler<dim, spacedim>::cell_iterator &)>;
```
I suspect that the fe is to be used on the cell; so the fe is *subordinate* to the cell. This would suggest that one should reverse the order of arguments, unless that leads to an incompatible change.

marcfehling replied:
Agreed! The `FiniteElement` is the future finite element that will be associated to the cell after repartitioning.

In fact, we never use the `cell_iterator` in all of these weighting functions, but I wanted to leave it in case someone comes up with some fancy `weighting_function` on his own.

bangerth commented on include/deal.II/distributed/cell_weights.h:
```diff
@@ -38,80 +38,178 @@ namespace parallel
+     * \mathrm{truncate}\left( a \, n_K^b \right) \f]
+     */
+    static weighting_function
+    ndofs_exponential_weighting(std::pair<double, double> coefficients);
```
```suggestion
    ndofs_exponential_weighting(const std::pair<double, double> &coefficients);
```

bangerth commented on include/deal.II/distributed/cell_weights.h:
```diff
@@ -38,80 +38,178 @@ namespace parallel
+      const typename Triangulation<dim, spacedim>::cell_iterator &cell,
+      const typename Triangulation<dim, spacedim>::CellStatus     status)>
+    make_weighting_callback(const hp::DoFHandler<dim, spacedim> &dof_handler,
+                            const weighting_function weighting_function);
```
```suggestion
                            const weighting_function &weighting_function);
```
Same elsewhere

bangerth commented on include/deal.II/distributed/cell_weights.h:
```diff
@@ -38,80 +38,178 @@ namespace parallel
+     */
+    static weighting_function
+    ndofs_exponential_weighting(
+      std::vector<std::pair<double, double>> coefficients);
```
```suggestion
      const std::vector<std::pair<double, double>> &coefficients);
```

bangerth commented on include/deal.II/distributed/cell_weights.h:
```diff
@@ -147,40 +255,33 @@ namespace parallel
+      const typename Triangulation<dim, spacedim>::CellStatus     status,
+      const hp::DoFHandler<dim, spacedim> &                       dof_handler,
+      const parallel::TriangulationBase<dim, spacedim> &          triangulation,
+      const weighting_function weighting_function);
```
```suggestion
      const weighting_function &weighting_function);
```

bangerth commented:
On 9/3/19 6:28 AM, marcfehling wrote:
> Shall we leave it in or remove it?

Leave it.

marcfehling commented:
Would it make sense to also consider `hp::QCollection` objects in the weighting process?

This would involve major changes to the interface of this class, and there may be no demand for this feature at all. I would leave it out for now and probably come back to that idea once someone requests it.

bangerth commented:
Yes, one thing at a time!

marcfehling commented:
I would like to move the static member functions from the `CellWeights` class into a separate namespace `CellWeightingFunctions`. The current interface looks rather ugly with these `template` specifiers and I think this would clean up things at least a bit:
```
parallel::CellWeights<dim, spacedim> cell_weights(
  hp_dof_handler,
  parallel::CellWeights<dim, spacedim>::template ndofs_weighting<
    unsigned int, unsigned int>({1000, 1}));
```

marcfehling commented:
Further: Would it make sense to provide functors instead of function pointers here? This should give us a performance boost.

marcfehling commented:
Some remarks on my previous comments:
- Introducing a new namespace does not yield to any benefit: We still need the `dim` and `spacedim` templates. We may rely on class template argument deduction introduced in `C++17` to clean out these lines :)
- Functors would work with `boost::signals`, but I think I'm over-engineering this. I'd be fine to stick to the function pointers for now.

bangerth commented on doc/news/changes/incompatibilities/20190903Fehling:
```diff
@@ -0,0 +1,3 @@
+Changed: The interface to the parallel::CellWeights class was changed.
```
It might be useful to be a little bit more explicit.

bangerth commented on include/deal.II/distributed/cell_weights.h:
```diff
@@ -38,80 +38,178 @@ namespace parallel
+   * CellWeights::reinit() function. The following code snippet demonstrates how
+   * to achieve each cell being weighted by its current number of degrees of
+   * freedom. We chose a factor of `1000` that corresponds to the initial weight
+   * each cell is assigned to upon creation.
```
Don't we mention this factor of 1000 somewhere in the documentation of the triangulation? I think it would be good to reference that place here to explain that factor of 1000.

marcfehling replied:
We actually do mention the section you are referring to in lines 74,75 (see https://github.com/dealii/dealii/pull/8687/commits/1d5d6132a74be88e4badc728d5481f2df9ef3635).

Do you think this is sufficient?

bangerth commented on include/deal.II/distributed/cell_weights.h:
```diff
@@ -38,80 +38,178 @@ namespace parallel
    * @code
-   * parallel::CellWeights<dim, spacedim> cell_weights(hp_dof_handler);
-   * cell_weights.register_ndofs_weighting();
+   * parallel::CellWeights<dim, spacedim> cell_weights(
```
could it be `const`?

marcfehling replied:
Unfortunately not.

`register_ndofs_weighting()` is a non-const function and changes class members, i.e. `connection`.

bangerth commented on include/deal.II/distributed/cell_weights.h:
```diff
@@ -38,80 +38,178 @@ namespace parallel
-   *   -> unsigned int {
-   *   return 1000 * std::pow(active_fe.dofs_per_cell, 1.6);
-   * });
+   * boost::signals2::connection listener =
```
```suggestion
   * boost::signals2::connection connection =
```

bangerth commented on include/deal.II/distributed/cell_weights.h:
```diff
@@ -38,80 +38,178 @@ namespace parallel
+   * weighting and load balancing.
    *
    * @note Be aware that this class connects the weight function to the
    * Triangulation during its construction. If the Triangulation
```
```suggestion
   * Triangulation during this class's constructor. If the Triangulation
```

bangerth commented on include/deal.II/distributed/cell_weights.h:
```diff
@@ -38,80 +38,178 @@ namespace parallel
+    constant_weighting(const unsigned int factor = 1000);
+
+    /**
+     * The pair of floating point numbers \f$(a,b)\f$ provided via
```
Here and below, remove the `\f` -- we have a script that allows us to write just `$...$`.
```suggestion
     * The pair of floating point numbers $(a,b)$ provided via
```

bangerth commented on include/deal.II/distributed/cell_weights.h:
```diff
@@ -38,80 +38,178 @@ namespace parallel
+     * The pair of floating point numbers \f$(a,b)\f$ provided via
+     * @p coefficients determines the weight \f$w_K\f$ of each cell \f$K\f$ with
+     * \f$n_K\f$ degrees of freedom in the following way: \f[ w_K =
+     * \mathrm{truncate}\left( a \, n_K^b \right) \f]
```
can you define what the truncate function actually does? I think you just want to omit that term in the formula and instead simply say "where the right hand side is rounded to the nearest integer since cell weights must be integers".

bangerth commented on include/deal.II/distributed/cell_weights.h:
```diff
@@ -38,80 +38,178 @@ namespace parallel
+     */
+    template <typename FactorType, typename ExponentType>
+    static WeightingFunction
+    ndofs_weighting(const std::pair<FactorType, ExponentType> &coefficients);
```
I looked at the code example above, and it's awkward to read because of the template arguments. Why not just let both of these be floating point numbers?

marcfehling replied:
We could let them be just floating point numbers. I thought we'd get some performance benefits from this since we'll work with `std::pow` on `floats`, but realize now that it would be negligibly small.

I'll remove the template parameters so we'll stick to floats.

bangerth replied:
Yes, that sounds good. I think the performance benefits are not going to be noticeable.

bangerth commented on include/deal.II/distributed/cell_weights.h:
```diff
@@ -38,80 +38,178 @@ namespace parallel
+    template <typename FactorType, typename ExponentType>
+    static WeightingFunction
+    ndofs_weighting(
+      const std::vector<std::pair<FactorType, ExponentType>> &coefficients);
```
same issues here

bangerth commented on include/deal.II/distributed/cell_weights.h:
```diff
@@ -147,40 +255,33 @@ namespace parallel
-     * returns the weight for each cell, determined by the weighting_function
-     * member.
+     * A callback function that will be connected to the cell_weight signal of
+     * the @p triangulation, which is attached to the @p dof_handler. Ultimately
```
```suggestion
     * the @p triangulation, to which the @p dof_handler is attached. Ultimately
```

bangerth commented on source/distributed/cell_weights.cc:
```diff
@@ -63,71 +44,186 @@ namespace parallel
-      -> unsigned int {
-      return factor * active_fe.dofs_per_cell * active_fe.dofs_per_cell;
-    };
+    if (std::is_same<ExponentType, unsigned int>::value)
```
you could save yourself some work by just picking one type for `ExponentType` :-)

bangerth commented on source/distributed/cell_weights.cc:
```diff
@@ -63,71 +44,186 @@ namespace parallel
+          result = std::trunc(result);
+
+          Assert(
+            result <= std::numeric_limits<unsigned int>::max(),
```
`results` also needs to be non-negative, by the way!

bangerth commented on source/distributed/cell_weights.cc:
```diff
@@ -63,71 +44,186 @@ namespace parallel
+      ExcMessage(
+        "parallel::CellWeights requires a parallel::TriangulationBase object."));
+
+    return std::bind(&CellWeights<dim, spacedim>::weighting_callback,
```
@masterleinad has been working hard to banish `std::bind` from the code base by using lambda functions instead -- he's going to stop liking you if you introduce new ones ;-)

bangerth commented on source/distributed/cell_weights.cc:
```diff
@@ -171,7 +267,113 @@ namespace parallel
+  {
+    triangulation =
+      dynamic_cast<const parallel::TriangulationBase<dim, spacedim> *>(
+        &(this->dof_handler->get_triangulation()));
```
Couldn't you move this into the member initializer list above already?

bangerth commented on source/distributed/cell_weights.cc:
```diff
@@ -171,7 +267,113 @@ namespace parallel
+    tria_listener.disconnect();
+
+    tria_listener = triangulation->signals.cell_weight.connect(
+      std::bind(&CellWeights<dim, spacedim>::weighting_callback,
```
same here as above

bangerth commented on source/distributed/cell_weights.cc:
```diff
@@ -171,7 +267,113 @@ namespace parallel
+    tria_listener.disconnect();
+
+    tria_listener = triangulation->signals.cell_weight.connect(
+      std::bind(&CellWeights<dim, spacedim>::weighting_callback,
```
and here

bangerth commented on source/distributed/cell_weights.cc:
```diff
@@ -171,7 +267,113 @@ namespace parallel
+    tria_listener.disconnect();
+
+    tria_listener = triangulation->signals.cell_weight.connect(
+      std::bind(&CellWeights<dim, spacedim>::weighting_callback,
```
and here

bangerth commented on source/distributed/cell_weights.cc:
```diff
@@ -171,7 +267,113 @@ namespace parallel
+      converted_function =
+        [&custom_function](
+          const typename hp::DoFHandler<dim, spacedim>::cell_iterator &cell,
+          const FiniteElement<dim, spacedim> &future_fe) -> unsigned int {
```
I don't think you actually need to specify the return type explicitly. It doesn't hurt, but also doesn't help.

marcfehling replied:
I'd prefer to be explicit about this :-)

bangerth replied:
I appreciate this -- I think we share that trait ;-)

bangerth commented on source/distributed/cell_weights.cc:
```diff
@@ -171,7 +267,113 @@ namespace parallel
+    };
+
+    tria_listener = triangulation->signals.cell_weight.connect(
+      std::bind(&CellWeights<dim, spacedim>::weighting_callback,
```
and here

marcfehling commented:
Addressed all suggestions. Once you reviewed these changes, I'll squash them together as usual (which will include rebasing on the master branch because of some merge conflicts).

bangerth commented:
Looks good, please squash.

/rebuild

marcfehling commented:
Testers have passed. Commits are squashed. Ready for merge.
